scripts/gulpfiles/docs_tasks.js

```diff
@@ -0,0 +1,129 @@
+ * This is used to create the `alternate_paths` data in the TOC.
+ * All subpages must be linked to their top-level page in the TOC for the
+ * left nav bar to remain correct after drilling down into a sub-page.
+ * @param {string[]} allFiles All files in docs directory.
+ * @returns {Map<string, string[]>}
+ */
+const buildAlternatePathsMap = function(allFiles) {
+    let map = new Map();
+    for (let file of allFiles) {
+        // Get the name of the class/namespaces/variable/etc., i.e. the top-level page.
+        let filePieces = file.split('.');
+        let name = filePieces[1];
+        if (!map.has(name)) {
+            map.set(name, []);
+        }
+        if (filePieces[2] !== 'md') {
+            // Don't add the top-level page to the map, only the sub-pages.
+            // Add all sub-pages to the array for this page.
+            map.get(name).push(file);
+        }
+    }
+    return map;
+}
+
+/**
+ * Create the _toc.yaml file used by devsite to create the leftnav.
+ * This file is generated from the contents of `blockly.md` which contains links
+ * to the other top-level API pages (each class, namespace, etc.).
+ */
+const createToc = function(done) {
+    const fileContent = fs.readFileSync(`${docsDir}/blockly.md`, 'utf8');
+    const files = fs.readdirSync(docsDir);
+    const map = buildAlternatePathsMap(files);
+    const referencePath = '/blockly/reference/js';
+    let content = 'toc:\n';
+
+    // Generate a section of TOC for each section/heading in the overview file.
+    const sections = fileContent.split('##');
+    for (let section of sections) {
+        // This converts the md table in each section to a JS object
+        const table = Extractor.extractObject(section, 'rows', false);
+        if (!table) {
+            continue;
+        }
+        // Get the name of the section, i.e. the text immediately after the `##` in the source doc
+        const sectionName = section.split('\n')[0].trim();
+        content += `- heading: ${sectionName}\n`
+        for (let row in table) {
+            // After going through the Extractor, the markdown is now HTML.
+            // Each row in the table is now a link (anchor tag).
+            // Get the target of the link, excluding the first `.` since we don't want a relative path.
+            const path = /href="\.(.*?)"/.exec(row)?.[1];
+            // Get the name of the link (text in between the <a> and </a>)
+            const name = /">(.*?)</.exec(row)?.[1];
+            if (!path || !name) {
+                continue;
+            }
+            content += `- title: ${name}\n`;
+            content += `  path: ${referencePath}${path}\n`;
+            // Get the list of sub-pages for this page.
+            let pages = map.get(path.split('.')[1]);
+            if (pages?.length) {
+                content += `  alternate_paths:\n`;
+                for (let page of pages) {
+                    content += `  - ${referencePath}/${page}\n`;
+                }
```

Comment:
```suggestion
                content += pages.map(page => `  - ${referencePath}/${page}\n`).join('');
```

Reply:
I think this is a personal preference thing, and the map isn't so superior to the for loop that it must be changed. I think it is slightly harder to read, so I left it the way it was.